add format checking of param values (#322)

momori256 opened the pull request:
Resolves #322

Adds the validation of param values so that they follow the syslog RFC.
The RFC requires param values to be a valid UTF-8 string.

The UTF-8 validation function is `validate_utf8_compliance` in src/validate.c. It is called by
`stumpless_new_param` and `stumpless_set_param_value` in src/param.c.

momori256 committed:
add format checking of param values (#322)

Adds the validation of param values so that they follow the
syslog RFC. The RFC requires param values to be a valid UTF-8 string.

goatshriek commented:
I don't have time for a full review right now, but at a glance this seems like it should be good to go! I will give it a full review in the next 24 hours and either request changes or merge it in.

momori256 committed:
remove 5 and 6 characters cases in UTF-8 string parsing

Since RFC 3629, only characters between U+0000 and U+10FFFF can be
encoded into UTF-8, so the maximum byte size of a valid UTF-8 string is
4 bytes.

momori256 committed:
update SetValue_ValidUTF8 test case to increase code coverage

Update SetValue_ValidUTF8 test to cover cases of strings with BOM and
strings encoded to 4 bytes.

goatshriek reviewed (approved):
Thank you very much for putting this together, as well as updating the tests for coverage. Excellent work!

momori256 commented:
Thank you for reviewing and merging!
